File: docs/build-log/followup-s2a-exam-scope/proof/guards.py
from pathlib import Path
import subprocess
import re

root = Path.cwd()
out = root / 'docs/build-log/followup-s2a-exam-scope/proof/guards'
out.mkdir(exist_ok=True)
endpoint = 'mcp/src/clinical-graph/exam-overview-endpoint.ts'
projection = 'mcp/src/clinical-graph/exam-overview-projection.ts'
store = 'mcp/src/clinical-graph/exam-scope-store.ts'
board = 'ui/src/components/charting/ExamOverviewBoard.tsx'
scene = 'ui/src/scenes/EncounterCharting.tsx'
server = ['node', '--import', 'tsx', '--test', 'mcp/tests/examOverviewEndpoint.test.ts', 'mcp/tests/examOverviewProjection.test.ts', 'mcp/tests/examScopeStore.test.ts']
ui = ['node', '--import', 'tsx', '--test', 'ui/tests/examOverviewBoard.test.tsx']
category_import = 'import { resolveVisitTypeCategoryForEncounter } from "../clinic/clinic-summary.js";\n'
category_change = ('examScope: scope.examScope,', 'examScope: (await resolveVisitTypeCategoryForEncounter(encounter, undefined, serviceFhir)) === "exams" ? "comprehensive" : "unknown",')
mutations = [
 ('G1', endpoint, [category_change], server, category_import),
 ('G2', endpoint, [('new FhirEncounterExamScopeStore(serviceFhir).get(encounterId)', 'Promise.resolve({ examScope: "comprehensive" })')], server, ''),
 ('G3', store, [('resource ? parseScope(resource) : { examScope: "comprehensive" }', 'resource ? parseScope(resource) : { examScope: "office-visit" }')], server, ''),
 ('G4', board, [('|| traceRows.length > 0 || groups.length > 0', '|| traceRows.length > 0')], ui, ''),
 ('G5', scene, [('onExamScopeChanged={refreshExamOverview}', 'onExamScopeChanged={() => { setVisitCharge(undefined); refreshExamOverview(); }}')], ui, ''),
 ('G6-row', board, [("Not required for this exam's scope", 'Not required for this visit type')], ui, ''),
 ('G6-trace', board, [("This count is relative to this exam's scope and is not a billing-code check.", 'This count is relative to the visit type and is not a billing-code check.')], ui, ''),
 ('G7', endpoint, [category_change], server, category_import),
 ('G8', scene, [('{!activeExamOverviewProjection && (', '{!activeExamOverviewProjection && examOverviewProjection?.examScope === "comprehensive" && (')], ui, ''),
 ('store-version', store, [('if ((existing?.meta?.versionId ?? null) !== expectedVersion) throw concurrentEdit();', '')], server, ''),
 ('store-create', store, [('if (result.examScope !== examScope || result.setAt !== setAt || result.setBy?.reference !== setBy.reference) throw concurrentEdit();', '')], server, ''),
]
def run(command, target):
 result = subprocess.run(command, capture_output=True, text=True)
 lines = (result.stdout + result.stderr).splitlines()
 summary = ['Command: ' + ' '.join(command), 'Exit: ' + str(result.returncode)]
 summary += [line.replace(str(root), '.') for line in lines if re.match(r'^(not ok |# (tests|pass|fail|cancelled|skipped|todo) )', line)]
 target.write_text('\n'.join(summary) + '\n')
 return result.returncode
for name, filename, replacements, command, prefix in mutations:
 path = root / filename
 original = path.read_text()
 mutated = original
 for before, after in replacements:
  assert before in mutated, (name, before)
  mutated = mutated.replace(before, after, 1)
 try:
  path.write_text(prefix + mutated)
  assert run(command, out / f'{name}-red.txt') != 0, name + ' survived mutation'
 finally:
  path.write_text(original)
 assert run(command, out / f'{name}-green.txt') == 0, name + ' failed after restoration'
 print(name + ': red then restored green', flush=True)

File: docs/build-log/followup-s2b1-shelf/proof/guards.mjs
import assert from 'node:assert/strict';
import { readFileSync, writeFileSync, mkdtempSync } from 'node:fs';
import { tmpdir } from 'node:os';
import { resolve, join } from 'node:path';
import { spawnSync } from 'node:child_process';

const root = resolve('.');
const logs = mkdtempSync(join(tmpdir(), 'odos-s2b1-guards-'));
const board = 'ui/src/components/charting/ExamOverviewBoard.tsx';
const scene = 'ui/src/scenes/EncounterCharting.tsx';
const endpoint = 'mcp/src/clinical-graph/finding-section-group-endpoint.ts';
const guards = [
  ['G1', board, 'shelfDefinitions.map(({ key, label }) => {', 'shelfDefinitions.filter(group => group.key !== "contact-lenses").map(({ key, label }) => {', 'ui', 'tests/examShelf.test.tsx'],
  ['G2', board, '<nav className="odos-exam-shelf"', '<details data-testid="chart-another-finding"><summary>Chart another finding</summary></details><nav className="odos-exam-shelf"', 'ui', 'tests/examShelf.test.tsx'],
  ['G3', scene, 'function openBoardEditor(sectionId: ChartSectionId) {', 'function openBoardEditor(sectionId: ChartSectionId) {\n    void fetch(`${clinicalGraphApiBase()}/clinical-graph/encounters/${encounterId}/exam-scope`, { method: "PUT", headers: { ...authHeaders(), "Content-Type": "application/json" }, body: JSON.stringify({ scope: "comprehensive" }) });', 'ui', 'tests/examOverviewBoard.test.tsx'],
  ['G4', board, 'availableSectionGroups.map(group => (', 'availableSectionGroups.slice(0, 0).map(group => (', 'ui', 'tests/examOverviewBoard.test.tsx'],
  ['G5', endpoint, 'if (sectionKeys.length) return { status: 409, body: { code: "section-group-has-content", sectionKeys } };', 'if (false && sectionKeys.length) return { status: 409, body: { code: "section-group-has-content", sectionKeys } };', 'mcp', 'tests/findingSectionGroup.test.ts'],
  ['G5-pin', board, 'Has findings this visit', 'Pinned', 'ui', 'tests/examShelf.test.tsx'],
  ['G6', board, 'const editor = editorForFinding(group, editorEntries);', 'const editor = editorForFinding(group, editorEntries);\n    if (!editor) continue;', 'ui', 'tests/examShelf.test.tsx'],
  ['G7', board, 'snapshotComponentCode(finding.current, "REFRACTION_TYPE") !== "FINAL_RX"', 'finding.provenance.state === "current" && snapshotComponentCode(finding.current, "REFRACTION_TYPE") !== "FINAL_RX"', 'ui', 'tests/examShelf.test.tsx'],
  ['G8', board, 'traceRows.length > 0 || (definition.optional', '(projection as ExamOverviewProjection & { visitTypeCategoryId?: string }).visitTypeCategoryId === "exams" || (definition.optional', 'ui', 'tests/examShelf.test.tsx'],
];
const selected = process.argv[2];
const outputPath = join(root, 'docs/build-log/followup-s2b1-shelf/guards.json');
const results = selected ? JSON.parse(readFileSync(outputPath, 'utf8')).results.filter(result => result.guard !== selected) : [];
function run(guard, state, pkg, file) {
  const args = ['--import', 'tsx', '--test', file];
  const result = spawnSync(process.execPath, args, { cwd: join(root, pkg), encoding: 'utf8' });
  const output = `${result.stdout ?? ''}${result.stderr ?? ''}`;
  writeFileSync(join(logs, `${guard}-${state}.log`), output);
  const summary = output.split('\n').filter(line => /^(not ok |# (tests|pass|fail|skipped|cancelled) )/.test(line));
  assert.ok(summary.some(line => line.startsWith('# tests ')), `${guard} did not execute the test runner`);
  assert.ok(state === 'red' ? result.status !== 0 && summary.some(line => line.startsWith('not ok ')) : result.status === 0, `${guard} ${state} did not produce expected result`);
  console.log(`${guard} ${state}: ${summary.join('; ')}`);
  return { exit: result.status, output: summary };
}
for (const [guard, file, before, after, pkg, test] of guards) {
  if (selected && selected !== guard) continue;
  const path = join(root, file);
  const original = readFileSync(path, 'utf8');
  assert.equal(original.split(before).length, 2, `${guard} mutation anchor must be unique`);
  let red;
  try {
    writeFileSync(path, original.replace(before, after));
    red = run(guard, 'red', pkg, test);
  } finally { writeFileSync(path, original); }
  const green = run(guard, 'green', pkg, test);
  assert.equal(readFileSync(path, 'utf8'), original);
  results.push({ guard, file, mutation: after, command: `cd ${pkg} && node --import tsx --test ${test}`, red, green });
}
const diff = spawnSync('git', ['diff', '--stat', '--', endpoint], { cwd: root, encoding: 'utf8' });
assert.equal(diff.status, 0); assert.equal(diff.stdout, '');
writeFileSync(outputPath, JSON.stringify({ results, endpointDiffStat: '' }, null, 2) + '\n');
console.log('G1–G8 restored; endpoint git diff --stat output is empty.');
